### ISSUE TITLE
[Inductor][CPU] Preallocated indexed-update workspace still allocates a fixed temporary per call

### ISSUE THREAD

**Issue by ruddyscent** [pytorch/pytorch#195330]
## Human commentary

I am continuing the upstream investigation for the GMES Torch-only CPU allocation gate. I asked Codex to separate the remaining compact indexed-update allocation from the full-buffer clone in #195285, validate it on current main, inspect the generated code, and search for duplicates. I reviewed the narrowed result before submitting it.

AI assistance disclosure: Codex prepared and executed the reproducer, inspected the generated wrapper and C++ kernel, searched related issues, and drafted the technical analysis below.

## Performance issue

A CPU compiled function that computes indexed values from a tensor before mutating that same tensor needs snapshot semantics when runtime indices may repeat. Inductor therefore creates a fixed-size temporary, which is reasonable for the direct data-dependent form.

However, the same allocation remains when the caller supplies an explicit fixed-shape workspace and computes the values into that workspace before the indexed write. The generated kernel allocates another same-sized buffer, then writes the result to both the tensor and the supplied workspace. The explicit workspace does not prevent allocator churn.

## Minimal reproducer

    import collections
    import json
    from pathlib import Path

    import torch

    torch.set_num_threads(1)
    torch.set_num_interop_threads(1)

    N = 51_520
    K = 99
    indices = torch.arange(0, 2 * K, 2, dtype=torch.int64)
    gain = torch.linspace(0.99, 1.01, K, dtype=torch.float64)
    values = torch.linspace(0.1, 0.2, K, dtype=torch.float64)
    base = torch.linspace(0.25, 1.25, N, dtype=torch.float64)

    def independent(field, indices, values):
        field.index_copy_(0, indices, values)

    def dependent(field, indices, gain):
        values = torch.index_select(field, 0, indices) * gain
        field.index_copy_(0, indices, values)

    def preallocated(field, indices, gain, workspace):
        torch.index_select(field, 0, indices, out=workspace)
        workspace.mul_(gain)
        field.index_copy_(0, indices, workspace)

    def allocation_histogram(name, function, arguments):
        compiled = torch.compile(
            function, backend="inductor", fullgraph=True, dynamic=False
        )
        field = base.clone()
        compiled(field, *arguments)
        compiled(field, *arguments)

        scope_name = "indexed-workspace::" + name
        trace_path = Path("/tmp/" + name + ".json")
        with torch.profiler.profile(
            activities=[torch.profiler.ProfilerActivity.CPU],
            profile_memory=True,
        ) as profile:
            for _ in range(5):
                with torch.profiler.record_function(scope_name):
                    compiled(field, *arguments)
        profile.export_chrome_trace(str(trace_path))

        events = json.loads(trace_path.read_text()).get("traceEvents", ())
        scopes = [
            event
            for event in events
            if event.get("name") == scope_name and event.get("ph") == "X"
        ]
        sizes = [
            int(event["args"]["Bytes"])
            for event in events
            if event.get("name") == "[memory]"
            and int(event.get("args", {}).get("Bytes", 0)) > 0
            and any(
                event.get("pid") == scope.get("pid")
                and event.get("tid") == scope.get("tid")
                and scope["ts"]
                <= event.get("ts", -1)
                <= scope["ts"] + scope["dur"]
                for scope in scopes
            )
        ]
        return dict(sorted(collections.Counter(sizes).items()))

    workspace = torch.empty(K, dtype=torch.float64)
    print(
        {
            "independent": allocation_histogram(
                "independent", independent, (indices, values)
            ),
            "dependent": allocation_histogram(
                "dependent", dependent, (indices, gain)
            ),
            "preallocated": allocation_histogram(
                "preallocated", preallocated, (indices, gain, workspace)
            ),
        }
    )

Current main result:

    {
        "independent": {},
        "dependent": {792: 5},
        "preallocated": {792: 5},
    }

The result is exact across repeated eager and compiled calls. The histogram maps allocation bytes to event count, so both data-dependent forms allocate one 99-element float64 buffer on every call.

## Generated-code diagnosis

For the preallocated case, the generated wrapper contains:

    buf0 = empty_strided_cpu((99,), (1,), torch.float64)
    cpp_fused_copy__index_copy_index_select_mul_0(
        indices, field, gain, buf0, field, workspace
    )

The generated C++ uses two loops. The first computes the 99 indexed values into buf0. The second reads buf0, scatters into field, and also stores the same values into workspace.

The temporary snapshot is semantically necessary for arbitrary runtime indices if values are read from and written to the same field, because duplicate indices can introduce read-after-write changes. This report does not request unsafe unconditional loop fusion.

The narrower missed optimization is that the preallocated form already provides storage for that snapshot. Inductor should use workspace as the intermediate and then scatter from it, rather than allocate buf0 and later copy the same values into workspace.

## Expected behavior

After warm-up, the preallocated form should preserve exact mutation and workspace results with zero per-call allocation. A fix should retain the temporary for the direct dependent form unless uniqueness or another safety condition is proven.

Useful regression boundaries are:

- independent values: zero allocation;
- dependent values without workspace: temporary retained when required for semantics;
- dependent values with explicit workspace: zero allocation;
- duplicate runtime indices: eager and compiled results remain equal;
- workspace aliasing field or indices: reject reinplacement or retain a safe temporary.

## Relationship to related issues

- #195285 and draft PR #195292 address a full-field generalized-scatter clone before indexed copy-back. The follow-up at ruddyscent/pytorch@be4d471 removes that full-field clone but leaves this compact workspace allocation.
- #195329 concerns multiple view mutations plus a returned alias and a full-base allocation. This case returns nothing and isolates a fixed indexed-values workspace.
- In the motivating GMES production trace, current master already has no full-field allocation. The five-step CPU trace contains 740,330 allocated bytes across 530 events. The reinplace follow-up changes neither number nor the allocation histogram. The 792-byte bucket alone occurs 65 times over five steps.

## Versions

- Python compiler source: current main at e4d9b6187e6ef2635cc2b648fbb409d25d6a9d9a
- Native extension build: ABI-compatible CPU build from b3632e56519f39da86e50fb0aa27235d626e8306
- Reported torch version: 2.15.0a0+gite4d9b61
- Python 3.12.3
- Ubuntu 24.04.4, x86_64
- Intel Core i7-6700
- CPU only, float64, fullgraph=True, dynamic=False

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @bobrenjc93 @aorenste
